File: backend/src/main/java/com/horain/llm/ChatMessage.java
package com.horain.llm;

import java.util.List;

/**
 * A message in the chat conversation.
 * For tool-role messages, toolCallId must be set.
 */
public record ChatMessage(
        String role,
        String content,
        List<ToolCallRequest> toolCalls,
        String toolCallId
) {

    public static ChatMessage user(String content) {
        return new ChatMessage("user", content, null, null);
    }

    public static ChatMessage assistant(String content) {
        return new ChatMessage("assistant", content, null, null);
    }

    public static ChatMessage assistantWithToolCalls(String content, List<ToolCallRequest> toolCalls) {
        return new ChatMessage("assistant", content, toolCalls, null);
    }

    public static ChatMessage tool(String content, String toolCallId) {
        return new ChatMessage("tool", content, null, toolCallId);
    }

    public static ChatMessage system(String content) {
        return new ChatMessage("system", content, null, null);
    }
}

File: backend/src/main/kotlin/com/horain/llm/ChatMessage.kt
package com.horain.llm

import kotlin.jvm.JvmRecord

/**
 * A message in the chat conversation.
 * For tool-role messages, toolCallId must be set.
 */
@JvmRecord
data class ChatMessage(
    val role: String,
    val content: String?,
    val toolCalls: List<ToolCallRequest>?,
    val toolCallId: String?
) {
    companion object {
        @JvmStatic
        fun user(content: String) = ChatMessage("user", content, null, null)

        @JvmStatic
        fun assistant(content: String) = ChatMessage("assistant", content, null, null)

        @JvmStatic
        fun assistantWithToolCalls(content: String, toolCalls: List<ToolCallRequest>) =
            ChatMessage("assistant", content, toolCalls, null)

        @JvmStatic
        fun tool(content: String, toolCallId: String) =
            ChatMessage("tool", content, null, toolCallId)

        @JvmStatic
        fun system(content: String) = ChatMessage("system", content, null, null)
    }
}
